### PR TITLE
To pandas

### PR DESCRIPTION
**Pull request by deng113jie**
issue #201 to_pandas for dataframe

### CODE REVIEW

File: exetera/core/dataframe.py [hunk 1]
```diff
@@ -530,8 +530,31 @@ def to_csv(self, filepath:str, row_filter:Union[np.ndarray, fld.Field]=None, col
                 else:
                     start_row += chunk_row_size
 
-            
-    def drop_duplicates(self, by: Union[str, List[str]], 
+    def to_pandas(self, row_filter=None, col_filter: Union[str, List[str]] = None):
+        """
+        Convert an ExeTera dataframe to Pandas DataFrame.
+        :param row_filter:
```
Comment (ericspod) [inline exetera/core/dataframe.py:536]: Need to explain what these parameters are, could a type for `row_filter` be provided?

File: exetera/core/dataframe.py [hunk 2]
```diff
@@ -530,8 +530,31 @@ def to_csv(self, filepath:str, row_filter:Union[np.ndarray, fld.Field]=None, col
...
+        :returns: A pandas dataframe.
+
+        Example::
+
+            pandas_df = df.to_pandas()
+        """
+        col_to_convert = list(self._columns.keys()) if col_filter is None else col_filter
+        if isinstance(col_to_convert, list):  # checking data length if multiple columns
+            bench_length = len(self._columns[col_to_convert[0]].data)
+            for field in col_to_convert:
```
Comment (ericspod) [inline exetera/core/dataframe.py:547]: Could we instead take the minimum length of the columns and convert only that many? Or is having different length columns an error anyway?

File: exetera/core/dataframe.py [hunk 3]
```diff
@@ -530,8 +530,31 @@ def to_csv(self, filepath:str, row_filter:Union[np.ndarray, fld.Field]=None, col
...
+                if len(self._columns[field].data) != bench_length:
+                    raise ValueError("All fields must be of the same length.")
+
+        col_to_convert = col_to_convert if isinstance(col_to_convert, list) else [col_to_convert]  # case of one column
+        temp = {}
+        for field in col_to_convert:
+            temp[field] = self._columns[field].data[:] if row_filter is None else np.array(self._columns[field].data[:])[row_filter]
```
Comment (ericspod) [inline exetera/core/dataframe.py:554]: Can we not do `self._columns[field].data[row_filter]` directly?

File: tests/test_dataframe.py [hunk 4]
```diff
@@ -932,6 +933,99 @@ def test_to_csv_with_row_filter_field(self):
 
         os.close(fd_csv)      
 
+
+class TestdataFrameToPandas(unittest.TestCase):
```
Comment (ericspod) [inline tests/test_dataframe.py:937]: I think we can reduce the size of the test code down a lot with `setupUp/tearDown`:

```python
class TestdataFrameToPandas(unittest.TestCase):
    def setUp():
        numeric_data = [i for i in range(20)]
        fixed_str_data = [b'a' for i in range(20)]
        ts_data = [1632234128 + i for i in range(20)]
        categorical_data = [1 for i in range(20)]
        idx_str_data = ['abc' for i in range(20)]

        bio = BytesIO()
        self.s=session.Session()
        dst = self.s.open_dataset(bio, 'w', 'dst')
        self.df = dst.create_dataframe('df')
        self.df.create_numeric('num', 'int32').data.write(numeric_data)
        self.df.create_fixed_string('fs1', 1).data.write(fixed_str_data)
        self.df.create_timestamp('ts1').data.write(ts_data)
        self.df.create_categorical('c1', 'int32', {'a': 1, 'b': 2}).data.write(categorical_data)
        self.df.create_indexed_string('is1').data.write(idx_str_data)
        
    def tearDown():
        self.s.close()
            
    def test_to_pandas_df(self):
        pddf1 = self.df.to_pandas()
        pddf2 = pd.DataFrame({'num':numeric_data, 'fs1':fixed_str_data, 'ts1':ts_data, 'c1':categorical_data, 'is1':idx_str_data}).astype(pddf1.dtypes)
        self.assertTrue(pddf1.equals(pddf2))

    def test_to_pandas_df_row_filter(self):
        row_filter = [True if i%2==0 else False for i in range(20)]
        pddf1 = self.df.to_pandas(row_filter=row_filter)
        pddf2 = pd.DataFrame({'num': numeric_data, 'fs1': fixed_str_data, 'ts1': ts_data, 'c1': categorical_data,
                              'is1': idx_str_data}).astype(pddf1.dtypes).loc[row_filter].reset_index(drop=True)
        self.assertTrue(pddf1.equals(pddf2))

    def test_to_pandas_df_col_filter(self):
        pddf1 = self.df.to_pandas(col_filter='num')
        pddf2 = pd.DataFrame({'num': numeric_data}).astype(pddf1.dtypes)
        self.assertTrue(pddf1.equals(pddf2))

        pddf1 = self.df.to_pandas(col_filter=['num', 'fs1', 'ts1', 'c1'])
        pddf2 = pd.DataFrame({'num': numeric_data, 'fs1': fixed_str_data, 'ts1': ts_data, 'c1': categorical_data}).astype(pddf1.dtypes)
        self.assertTrue(pddf1.equals(pddf2))

    def test_to_pandas_df_filter(self):
        row_filter = [True if i % 2 == 0 else False for i in range(20)]
        col_filter = ['num', 'fs1', 'ts1', 'c1']
        pddf1 = self.df.to_pandas(row_filter=row_filter, col_filter=col_filter)
        pddf2 = pd.DataFrame({'num': numeric_data, 'fs1': fixed_str_data, 'ts1': ts_data,
                              'c1': categorical_data}).astype(pddf1.dtypes).loc[row_filter].reset_index(drop=True)
        self.assertTrue(pddf1.equals(pddf2))

```

File: exetera/core/dataframe.py (hunk 2 above)
Comment (deng113jie) [reply exetera/core/dataframe.py:547]: Yep,  pandas will raise an error for having different length.

File: exetera/core/dataframe.py (hunk 3 above)
Comment (deng113jie) [reply exetera/core/dataframe.py:554]: Cause the data[:] for indexed_string will return a list and won't support [row_filter]

File: tests/test_dataframe.py (hunk 4 above)
Comment (deng113jie) [reply tests/test_dataframe.py:937]: Yes, good point. Thanks!

File: exetera/core/dataframe.py (hunk 1 above)
Comment (deng113jie) [reply exetera/core/dataframe.py]: updated.

File: exetera/core/dataframe.py (hunk 3 above)
Comment (ericspod) [reply exetera/core/dataframe.py:554]: OK would this be too expensive then in the ase that `row_filter` is `None`?
```python
field_arr=np.array(self._columns[field].data[:])
temp[field] = field_arr if row_filter is None else field_arr[row_filter]
```

File: exetera/core/dataframe.py (hunk 2 above)
Comment (ericspod) [reply exetera/core/dataframe.py:547]: Line 549 is the only new line here missed by the unit tests according to the coverage report so could you provide a simple test for this failure case?

**Review by ericspod — APPROVED** [review]

**Pull request merged**